From 602ba288ef3ebef04eb8ccaa80c4f7e9bd49cb6e Mon Sep 17 00:00:00 2001
From: Grant Gardner <grant@lastweekend.com.au>
Date: Thu, 10 Feb 2011 23:48:01 +1100
Subject: [PATCH 1/3] allow nil for readder filler

---
 ext/filler.c |   13 +++++++++----
 1 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/ext/filler.c b/ext/filler.c
index 8372470..6cd93fb 100644
--- a/ext/filler.c
+++ b/ext/filler.c
@@ -16,10 +16,15 @@ VALUE rfiller_new(VALUE class){
 VALUE rfiller_push(VALUE self, VALUE name, VALUE stat, VALUE offset) {
   struct filler_t *f;
   Data_Get_Struct(self,struct filler_t,f);
-  struct stat st;
-  memset(&st, 0, sizeof(st));
-  rstat2stat(stat,&st);
-  f->filler(f->buffer,STR2CSTR(name),&st,NUM2LONG(offset));
+  //Allow nil return instead of a stat
+  if (NIL_P(stat)) {
+    f->filler(f->buffer,STR2CSTR(name),NULL,NUM2LONG(offset));
+  } else {
+    struct stat st;
+    memset(&st, 0, sizeof(st));
+    rstat2stat(stat,&st);
+    f->filler(f->buffer,STR2CSTR(name),&st,NUM2LONG(offset));
+  }
   return self;
 }
 
-- 
1.7.1

